mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+        )
+        return next_target_params
+
+    def soft_update(
+        next_online_params: FrozenVariableDict, target_params: FrozenVariableDict, t_train: int
+    ) -> FrozenVariableDict:
+        next_target_params = optax.incremental_update(
+            next_online_params, target_params, cfg.system.tau
+        )
+        return next_target_params
+
+    # TRAIN LEVEL 2
+    def update_q(
+        params: DDQNParams, opt_states: optax.OptState, data: Transition, t_train: int
+    ) -> Tuple[DDQNParams, optax.OptState, Metrics]:
+        """Update the Q parameters."""
+
+        # Get the data associated with obs
+        data_first = jax.tree_map(lambda x: x[:, :-1, ...], data)
+        # Get the data associated with next_obs
+        data_next = jax.tree_map(lambda x: x[:, 1:, ...], data)
+
+        first_reward = data_first.reward
+        next_done = data_next.done
+
+        # Scan over each sample and discard first timestep
+        hidden_state, obs_done = prep_inputs_to_scannedrnn(data.obs, data.done)
+
+        _, next_online_greedy_dist = q_net.apply(
+            params.online, hidden_state, obs_done
+        )  # eps defaults to 0
+
+        _, next_q_vals_target = q_net.apply(
+            params.target, hidden_state, obs_done, method="get_q_values"
+        )
+
+        # Skip first timestep, since we are working with "next" information
+        next_action = next_online_greedy_dist.mode()[1:, ...]  # (T, B, ...)
+        next_q_vals_target = next_q_vals_target[1:, ...]  # (T, B, ...)
+
+        # Double q-value selection
+        next_q_val = jnp.squeeze(
+            jnp.take_along_axis(next_q_vals_target, next_action[..., jnp.newaxis], axis=-1), axis=-1
+        )
+
+        next_q_val = switch_leading_axes(next_q_val)  # (T, B, ...) -> (B, T, ...)
+
+        # TD Target
+        target_q_val = (
+            first_reward
+            + (1.0 - jnp.array(next_done, dtype="float32")) * cfg.system.gamma * next_q_val
+        )
+
+        # Update Q function.
+        q_grad_fn = jax.grad(q_loss_fn, has_aux=True)
+        q_grads, q_loss_info = q_grad_fn(
+            params.online, data_first.obs, data_first.done, data_first.action, target_q_val
+        )
+
+        # Mean over the device and batch dimension.
+        q_grads, q_loss_info = lax.pmean((q_grads, q_loss_info), axis_name="device")
+        q_grads, q_loss_info = lax.pmean((q_grads, q_loss_info), axis_name="batch")
+        q_updates, next_opt_state = opt.update(q_grads, opt_states)
+        next_online_params = optax.apply_updates(params.online, q_updates)
+
+        target_update_params = (next_online_params, params.target, t_train)
+
+        # Depending on choice of target network strategy, update target weights.
+        next_target_params = jax.lax.cond(
+            cfg.system.hard_update,
+            hard_update,
+            soft_update,
+            *target_update_params,
+        )
```

Comment:
Isn't Q learning always a hard update?

If we're going to keep this option then could we rather remove the cond and use a normal python if, it'll compile out the branch we don't use since `cfg.system.hard_update` is known at compile time and you won't need those 2 functions anymore

Reply:
I'd like to keep it as an option, especially with the anakin architecture a soft update may not be a bad idea. As for the if statement, that makes sense, willdo!